File: packages/nerva-py/nerva/runtime/container.py
"""Container runtime — execute handlers in Docker containers (N-612).

Runs handler logic inside Docker containers via ``docker run``, passing
input as JSON on stdin and collecting output from stdout. Supports
resource limits, network isolation, timeout via container kill, and
per-handler circuit breakers.
"""

from __future__ import annotations

import asyncio
import json
import logging
import time
from dataclasses import asdict, dataclass, field

from nerva.context import ExecContext
from nerva.runtime import AgentInput, AgentResult, AgentStatus
from nerva.runtime.circuit_breaker import CircuitBreaker, CircuitBreakerConfig

__all__ = [
    "ContainerRuntime",
    "ContainerConfig",
    "ContainerHandlerConfig",
]

_log = logging.getLogger(__name__)

# -- Constants ---------------------------------------------------------------

DEFAULT_TIMEOUT_SECONDS: float = 60.0
"""Default timeout for container execution."""

DEFAULT_MEMORY_LIMIT: str = "256m"
"""Default Docker memory limit."""

DEFAULT_CPU_LIMIT: str = "1.0"
"""Default Docker CPU quota (number of CPUs)."""

DEFAULT_NETWORK_MODE: str = "none"
"""Default Docker network mode (isolated by default)."""

MAX_OUTPUT_BYTES: int = 1_048_576  # 1 MB
"""Maximum stdout bytes collected from a container."""


# -- Config ------------------------------------------------------------------


@dataclass(frozen=True)
class ContainerHandlerConfig:
    """Configuration for a single container-based handler.

    Attributes:
        image: Docker image name (e.g. ``"myorg/search:latest"``).
        memory_limit: Docker memory limit (e.g. ``"256m"``).
        cpu_limit: Docker CPU quota as a string (e.g. ``"0.5"``).
        network_mode: Docker network mode (e.g. ``"none"``, ``"bridge"``).
        env: Environment variables passed to the container.
    """

    image: str
    memory_limit: str = DEFAULT_MEMORY_LIMIT
    cpu_limit: str = DEFAULT_CPU_LIMIT
    network_mode: str = DEFAULT_NETWORK_MODE
    env: dict[str, str] = field(default_factory=dict)


@dataclass(frozen=True)
class ContainerConfig:
    """Global configuration for the container runtime.

    Attributes:
        timeout_seconds: Max execution time per container invocation.
        circuit_breaker: Circuit breaker thresholds applied per handler.
        max_output_bytes: Max stdout bytes to collect before truncation.
        docker_command: Path or name of the docker executable.
    """

    timeout_seconds: float = DEFAULT_TIMEOUT_SECONDS
    circuit_breaker: CircuitBreakerConfig | None = None
    max_output_bytes: int = MAX_OUTPUT_BYTES
    docker_command: str = "docker"


# -- Runtime -----------------------------------------------------------------


class ContainerRuntime:
    """Execute handlers inside Docker containers.

    Each handler maps to a Docker image. Input is serialized as JSON to
    the container's stdin; output is collected from stdout. Supports
    memory/CPU limits, network isolation, timeout via container kill,
    and per-handler circuit breakers.

    Args:
        config: Global runtime configuration. Uses defaults when None.
    """

    def __init__(self, config: ContainerConfig | None = None) -> None:
        self._config = config or ContainerConfig()
        self._handlers: dict[str, ContainerHandlerConfig] = {}
        self._breakers: dict[str, CircuitBreaker] = {}

    # -- Registration --------------------------------------------------------

    def register(self, name: str, handler_config: ContainerHandlerConfig) -> None:
        """Register a handler with its Docker container configuration.

        Args:
            name: Unique handler identifier.
            handler_config: Docker image and resource limits.

        Raises:
            ValueError: If *name* is empty or already registered.
            ValueError: If *handler_config.image* is empty.
        """
        if not name:
            raise ValueError("Handler name must not be empty")
        if name in self._handlers:
            raise ValueError(f"Handler '{name}' is already registered")
        if not handler_config.image or not handler_config.image.strip():
            raise ValueError("Container image must not be empty")

        self._handlers[name] = handler_config

    # -- Public API ----------------------------------------------------------

    async def invoke(
        self, handler: str, input: AgentInput, ctx: ExecContext
    ) -> AgentResult:
        """Run a handler inside a Docker container.

        Args:
            handler: Handler name (must be registered).
            input: Structured input serialized as JSON to stdin.
            ctx: Execution context for tracing and cancellation.

        Returns:
            AgentResult populated from container stdout, or an error result.
        """
        handler_cfg = self._handlers.get(handler)
        if handler_cfg is None:
            return _not_found_result(handler)

        breaker = self._get_breaker(handler)
        if not breaker.is_allowed():
            return _circuit_open_result(handler)

        ctx.add_event("container.start", handler=handler, image=handler_cfg.image)
        started_at = time.monotonic()

        result = await self._run_container(handler, handler_cfg, input, ctx)

        elapsed = time.monotonic() - started_at
        ctx.add_event(
            "container.end",
            handler=handler,
            status=result.status.value,
            elapsed_seconds=f"{elapsed:.3f}",
        )

        _record_on_breaker(breaker, result.status)
        return result

    async def invoke_chain(
        self, handlers: list[str], input: AgentInput, ctx: ExecContext
    ) -> AgentResult:
        """Run handlers in sequence, piping each output as the next input's message.

        Args:
            handlers: Ordered list of handler names.
            input: Initial input for the first handler.
            ctx: Execution context shared across the chain.

        Returns:
            AgentResult from the last successfully executed handler.

        Raises:
            ValueError: If *handlers* is empty.
        """
        if not handlers:
            raise ValueError("handlers list must not be empty")

        current_input = input
        result = AgentResult(status=AgentStatus.ERROR, handler="", error="no handlers ran")

        for handler_name in handlers:
            result = await self.invoke(handler_name, current_input, ctx)
            if result.status != AgentStatus.SUCCESS:
                return result
            current_input = AgentInput(
                message=result.output,
                args=current_input.args,
                tools=current_input.tools,
                history=current_input.history,
            )

        return result

    async def delegate(
        self, handler: str, input: AgentInput, parent_ctx: ExecContext
    ) -> AgentResult:
        """Invoke a handler from within another handler.

        Args:
            handler: Handler name to delegate to.
            input: Input for the delegated handler.
            parent_ctx: Parent's execution context.

        Returns:
            AgentResult from the delegated handler.
        """
        child_ctx = parent_ctx.child(handler)
        return await self.invoke(handler, input, child_ctx)

    # -- Private: container execution ----------------------------------------

    async def _run_container(
        self,
        handler: str,
        handler_cfg: ContainerHandlerConfig,
        input: AgentInput,
        ctx: ExecContext,
    ) -> AgentResult:
        """Spawn a Docker container, feed stdin, collect stdout.

        Args:
            handler: Handler name.
            handler_cfg: Docker configuration for this handler.
            input: Structured input to serialize as JSON.
            ctx: Execution context.

        Returns:
            AgentResult from parsed container output.
        """
        cmd = _build_docker_command(self._config.docker_command, handler_cfg)
        input_json = json.dumps(asdict(input))

        try:
            stdout, returncode = await self._spawn_and_collect(cmd, input_json)
        except asyncio.TimeoutError:
            return _timeout_result(handler, self._config.timeout_seconds)

        if returncode != 0:
            return _error_result_from_exit(handler, returncode, stdout)

        return _build_success_result(handler, stdout)

    async def _spawn_and_collect(
        self, cmd: list[str], input_json: str
    ) -> tuple[str, int]:
        """Spawn the docker process, write stdin, read stdout with timeout.

        Args:
            cmd: Full docker command with arguments.
            input_json: JSON payload to write to stdin.

        Returns:
            Tuple of (stdout_text, returncode).

        Raises:
            asyncio.TimeoutError: If the container exceeds the timeout.
        """
        process = await asyncio.create_subprocess_exec(
            *cmd,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        try:
            stdout_bytes, _ = await asyncio.wait_for(
                process.communicate(input=input_json.encode()),
                timeout=self._config.timeout_seconds,
            )
        except asyncio.TimeoutError:
            await _kill_process(process)
            raise

        # Truncate oversized output
        if len(stdout_bytes) > self._config.max_output_bytes:
            stdout_bytes = stdout_bytes[: self._config.max_output_bytes]

        returncode = process.returncode if process.returncode is not None else -1
        return stdout_bytes.decode(errors="replace"), returncode

    # -- Circuit breaker helpers ---------------------------------------------

    def _get_breaker(self, handler: str) -> CircuitBreaker:
        """Return the circuit breaker for *handler*, creating one if needed.

        Args:
            handler: Handler name used as the breaker key.

        Returns:
            The CircuitBreaker instance for this handler.
        """
        if handler not in self._breakers:
            self._breakers[handler] = CircuitBreaker(self._config.circuit_breaker)
        return self._breakers[handler]


# -- Pure helpers ------------------------------------------------------------


def _build_docker_command(
    docker_cmd: str, handler_cfg: ContainerHandlerConfig
) -> list[str]:
    """Build the full ``docker run`` command from configuration.

    Args:
        docker_cmd: Docker executable path.
        handler_cfg: Handler-specific container configuration.

    Returns:
        Command list ready for subprocess execution.
    """
    cmd = [
        docker_cmd, "run",
        "--rm",
        "-i",
        f"--memory={handler_cfg.memory_limit}",
        f"--cpus={handler_cfg.cpu_limit}",
        f"--network={handler_cfg.network_mode}",
    ]
    for key, value in handler_cfg.env.items():
        cmd.extend(["-e", f"{key}={value}"])
    cmd.append(handler_cfg.image)
    return cmd


def _record_on_breaker(breaker: CircuitBreaker, status: AgentStatus) -> None:
    """Record success or failure on the circuit breaker.

    Args:
        breaker: The handler's circuit breaker.
        status: The outcome status of the invocation.
    """
    if status == AgentStatus.SUCCESS:
        breaker.record_success()
    else:
        breaker.record_failure()


async def _kill_process(process: asyncio.subprocess.Process) -> None:
    """Terminate a subprocess and wait for it to exit.

    Args:
        process: The subprocess to kill.
    """
    try:
        process.kill()
    except ProcessLookupError:
        return
    await process.wait()


def _not_found_result(handler: str) -> AgentResult:
    """Build an error result for an unregistered handler.

    Args:
        handler: The handler name.

    Returns:
        AgentResult with ERROR status.
    """
    return AgentResult(
        status=AgentStatus.ERROR,
        handler=handler,
        error=f"handler '{handler}' is not registered",
    )


def _circuit_open_result(handler: str) -> AgentResult:
    """Build an error result for a handler whose circuit is open.

    Args:
        handler: The handler that was rejected.

    Returns:
        AgentResult with ERROR status.
    """
    return AgentResult(
        status=AgentStatus.ERROR,
        handler=handler,
        error=f"circuit open for handler '{handler}'",
    )


def _timeout_result(handler: str, timeout_seconds: float) -> AgentResult:
    """Build a TIMEOUT result.

    Args:
        handler: Handler name.
        timeout_seconds: The timeout that was exceeded.

    Returns:
        AgentResult with TIMEOUT status.
    """
    return AgentResult(
        status=AgentStatus.TIMEOUT,
        handler=handler,
        error=f"container '{handler}' timed out after {timeout_seconds}s",
    )


def _error_result_from_exit(handler: str, returncode: int, output: str) -> AgentResult:
    """Build an ERROR result from a non-zero exit code.

    Args:
        handler: Handler name.
        returncode: Non-zero exit code.
        output: Container stdout.

    Returns:
        AgentResult with ERROR status.
    """
    return AgentResult(
        status=AgentStatus.ERROR,
        handler=handler,
        output=output,
        error=f"container '{handler}' exited with code {returncode}",
    )


def _build_success_result(handler: str, output: str) -> AgentResult:
    """Build a SUCCESS result from container output.

    Attempts to parse JSON from the output for structured data.

    Args:
        handler: Handler name.
        output: Raw container stdout.

    Returns:
        AgentResult with SUCCESS status.
    """
    data: dict[str, str] = {}
    try:
        parsed = json.loads(output.strip())
        if isinstance(parsed, dict):
            data = parsed
    except (json.JSONDecodeError, ValueError):
        pass

    response_text = data.pop("output", "") or data.pop("response", "") or output
    return AgentResult(
        status=AgentStatus.SUCCESS,
        output=response_text,
        data=data,
        handler=handler,
    )

File: packages/nerva-js/src/runtime/container.ts
/**
 * Container runtime — execute handlers in Docker containers.
 *
 * Spawns a Docker container for each handler invocation, passing input
 * as JSON on stdin and collecting output from stdout. Supports configurable
 * images, resource limits, network isolation, timeout via container kill,
 * and per-handler circuit breakers.
 *
 * @module runtime/container
 */

import { spawn } from "node:child_process";
import type { ExecContext } from "../context.js";
import {
  AgentStatus,
  type AgentInput,
  type AgentResult,
  type AgentRuntime,
  createAgentInput,
  createAgentResult,
} from "./index.js";
import { CircuitBreaker, type CircuitBreakerConfig } from "./circuit-breaker.js";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Default timeout for container execution in milliseconds. */
const DEFAULT_TIMEOUT_MS = 60_000;

/** Maximum stdout bytes collected before truncation. */
const MAX_OUTPUT_BYTES = 1_048_576; // 1 MB

/** Default Docker network mode. */
const DEFAULT_NETWORK_MODE = "none";

/** Regex to locate a JSON object in noisy container output. */
const JSON_EXTRACT_PATTERN = /\{[^{}]*\}|\{.*\}/s;

// ---------------------------------------------------------------------------
// ContainerConfig
// ---------------------------------------------------------------------------

/**
 * Configuration for a single container handler.
 */
export interface ContainerHandlerConfig {
  /** Docker image to run (e.g. `"myregistry/handler:latest"`). */
  readonly image: string;
  /** Memory limit (e.g. `"256m"`). Undefined means no limit. */
  readonly memoryLimit?: string;
  /** CPU limit (e.g. `"0.5"`). Undefined means no limit. */
  readonly cpuLimit?: string;
  /** Network mode (e.g. `"none"`, `"bridge"`). Defaults to `"none"`. */
  readonly networkMode?: string;
  /** Extra environment variables passed to the container. */
  readonly env?: Readonly<Record<string, string>>;
}

/**
 * Configuration for the container runtime.
 */
export interface ContainerRuntimeConfig {
  /** Max execution time per handler invocation in milliseconds. */
  readonly timeoutMs: number;
  /** Circuit breaker thresholds applied per handler. */
  readonly circuitBreaker?: Partial<CircuitBreakerConfig> | undefined;
  /** Maximum stdout bytes to collect before truncation. */
  readonly maxOutputBytes: number;
}

// ---------------------------------------------------------------------------
// ContainerRuntime
// ---------------------------------------------------------------------------

/**
 * Execute handlers as Docker containers with lifecycle management.
 *
 * Each handler maps to a container image. On invocation, the runtime
 * spawns `docker run`, writes `AgentInput` JSON to stdin, and collects
 * `AgentResult`-shaped JSON from stdout. Timeout is enforced by killing
 * the container.
 *
 * @example
 * ```ts
 * const runtime = new ContainerRuntime();
 * runtime.register("search", { image: "search-handler:latest" });
 * const result = await runtime.invoke("search", agentInput, ctx);
 * ```
 */
export class ContainerRuntime implements AgentRuntime {
  private readonly _config: ContainerRuntimeConfig;
  private readonly _handlers = new Map<string, ContainerHandlerConfig>();
  private readonly _breakers = new Map<string, CircuitBreaker>();

  /**
   * @param config - Optional runtime configuration overrides.
   */
  constructor(config?: Partial<ContainerRuntimeConfig>) {
    this._config = {
      timeoutMs: config?.timeoutMs ?? DEFAULT_TIMEOUT_MS,
      circuitBreaker: config?.circuitBreaker,
      maxOutputBytes: config?.maxOutputBytes ?? MAX_OUTPUT_BYTES,
    };
  }

  /**
   * Register a handler with its container configuration.
   *
   * @param name - Unique handler identifier.
   * @param handlerConfig - Docker image and resource limits.
   * @throws {Error} If a handler with the same name is already registered.
   */
  register(name: string, handlerConfig: ContainerHandlerConfig): void {
    if (this._handlers.has(name)) {
      throw new Error(`Handler '${name}' is already registered`);
    }
    this._handlers.set(name, handlerConfig);
  }

  /**
   * Run a handler in a Docker container.
   *
   * @param handler - Handler name (must be registered).
   * @param input - Structured input serialised to JSON on stdin.
   * @param ctx - Execution context for tracing and streaming.
   * @returns AgentResult populated from the container's stdout JSON.
   */
  async invoke(handler: string, input: AgentInput, ctx: ExecContext): Promise<AgentResult> {
    const handlerConfig = this._handlers.get(handler);
    if (handlerConfig === undefined) {
      return createAgentResult(AgentStatus.ERROR, {
        handler,
        error: `handler '${handler}' not found`,
      });
    }

    const breaker = this._getBreaker(handler);
    if (!breaker.isAllowed()) {
      return buildCircuitOpenResult(handler);
    }

    ctx.addEvent("container.start", { handler, image: handlerConfig.image });
    const inputJson = JSON.stringify(input);
    const { stdout, timedOut, exitCode } = await this._runContainer(
      handlerConfig,
      inputJson,
      ctx,
    );

    const result = buildResult(handler, stdout, exitCode, timedOut, this._config.timeoutMs);
    recordOnBreaker(breaker, result.status);
    ctx.addEvent("container.end", { handler, status: result.status });
    return result;
  }

  /**
   * Run handlers in sequence, piping each output as the next input's message.
   *
   * @param handlers - Ordered list of handler names.
   * @param input - Initial input for the first handler.
   * @param ctx - Execution context shared across the chain.
   * @returns AgentResult from the last successfully executed handler.
   * @throws {Error} If `handlers` is empty.
   */
  async invokeChain(
    handlers: string[],
    input: AgentInput,
    ctx: ExecContext,
  ): Promise<AgentResult> {
    if (handlers.length === 0) {
      throw new Error("handlers list must not be empty");
    }

    let currentInput = input;
    let result = createAgentResult(AgentStatus.ERROR, { error: "no handlers ran" });

    for (const handlerName of handlers) {
      result = await this.invoke(handlerName, currentInput, ctx);
      if (result.status !== AgentStatus.SUCCESS) {
        return result;
      }
      currentInput = createAgentInput(result.output, {
        args: { ...currentInput.args },
        tools: currentInput.tools.map((t) => ({ ...t })),
        history: currentInput.history.map((h) => ({ ...h })),
      });
    }

    return result;
  }

  /**
   * Invoke a handler from within another handler (agent-to-agent delegation).
   *
   * @param handler - Handler name to delegate to.
   * @param input - Input for the delegated handler.
   * @param parentCtx - Parent's execution context.
   * @returns AgentResult from the delegated handler.
   */
  async delegate(
    handler: string,
    input: AgentInput,
    parentCtx: ExecContext,
  ): Promise<AgentResult> {
    const childCtx = parentCtx.child(handler);
    return this.invoke(handler, input, childCtx);
  }

  // -- Private: container execution -----------------------------------------

  /**
   * Spawn a Docker container, pipe input, and collect output.
   *
   * @param handlerConfig - Container image and resource configuration.
   * @param inputJson - Serialised AgentInput to write to stdin.
   * @param ctx - Execution context for streaming.
   * @returns stdout content, timeout flag, and exit code.
   */
  private _runContainer(
    handlerConfig: ContainerHandlerConfig,
    inputJson: string,
    ctx: ExecContext,
  ): Promise<{ stdout: string; timedOut: boolean; exitCode: number }> {
    return new Promise((resolvePromise) => {
      const args = buildDockerArgs(handlerConfig);
      const child = spawn("docker", args, { stdio: ["pipe", "pipe", "pipe"] });

      const chunks: Buffer[] = [];
      let totalBytes = 0;
      let timedOut = false;

      const timer = setTimeout(() => {
        timedOut = true;
        child.kill("SIGKILL");
      }, this._config.timeoutMs);

      if (child.stdout !== null) {
        child.stdout.on("data", (chunk: Buffer) => {
          const remaining = this._config.maxOutputBytes - totalBytes;
          if (remaining <= 0) return;

          const trimmed = remaining < chunk.length ? chunk.subarray(0, remaining) : chunk;
          chunks.push(trimmed);
          totalBytes += trimmed.length;

          if (ctx.stream !== null) {
            ctx.stream.push(trimmed.toString("utf-8")).catch(() => {
              // Best-effort streaming
            });
          }
        });
      }

      child.on("error", () => {
        clearTimeout(timer);
        resolvePromise({ stdout: "", timedOut: false, exitCode: -1 });
      });

      child.on("close", (code) => {
        clearTimeout(timer);
        if (timedOut) {
          resolvePromise({ stdout: "", timedOut: true, exitCode: -1 });
          return;
        }
        const stdout = Buffer.concat(chunks).toString("utf-8");
        resolvePromise({ stdout, timedOut: false, exitCode: code ?? -1 });
      });

      if (child.stdin !== null) {
        child.stdin.write(inputJson, "utf-8");
        child.stdin.end();
      }
    });
  }

  /**
   * Return the circuit breaker for the handler, creating one if needed.
   *
   * @param handler - Handler name used as the breaker key.
   * @returns The CircuitBreaker instance for this handler.
   */
  private _getBreaker(handler: string): CircuitBreaker {
    let breaker = this._breakers.get(handler);
    if (breaker === undefined) {
      breaker = new CircuitBreaker(this._config.circuitBreaker);
      this._breakers.set(handler, breaker);
    }
    return breaker;
  }
}

// ---------------------------------------------------------------------------
// Pure helpers
// ---------------------------------------------------------------------------

/**
 * Build the `docker run` argument list from handler configuration.
 *
 * @param config - Container handler configuration.
 * @returns Array of CLI arguments for `docker run`.
 */
function buildDockerArgs(config: ContainerHandlerConfig): string[] {
  const args = ["run", "--rm", "-i"];

  const networkMode = config.networkMode ?? DEFAULT_NETWORK_MODE;
  args.push("--network", networkMode);

  if (config.memoryLimit !== undefined) {
    args.push("--memory", config.memoryLimit);
  }

  if (config.cpuLimit !== undefined) {
    args.push("--cpus", config.cpuLimit);
  }

  if (config.env !== undefined) {
    for (const [key, value] of Object.entries(config.env)) {
      args.push("-e", `${key}=${value}`);
    }
  }

  args.push(config.image);
  return args;
}

/**
 * Build an AgentResult from container output and execution status.
 *
 * @param handler - Handler name.
 * @param stdout - Raw stdout from the container.
 * @param exitCode - Container exit code.
 * @param timedOut - Whether the container was killed due to timeout.
 * @param timeoutMs - Configured timeout value for error messages.
 * @returns Populated AgentResult.
 */
function buildResult(
  handler: string,
  stdout: string,
  exitCode: number,
  timedOut: boolean,
  timeoutMs: number,
): AgentResult {
  if (timedOut) {
    return createAgentResult(AgentStatus.TIMEOUT, {
      handler,
      error: `container '${handler}' timed out after ${timeoutMs}ms`,
    });
  }

  if (exitCode !== 0) {
    return createAgentResult(AgentStatus.ERROR, {
      handler,
      output: stdout,
      error: `container '${handler}' exited with code ${exitCode}`,
    });
  }

  const output = extractOutputText(stdout);
  return createAgentResult(AgentStatus.SUCCESS, {
    handler,
    output,
  });
}

/**
 * Extract the response text from container stdout.
 *
 * Tries to parse as JSON and pull an "output" or "response" field.
 * Falls back to raw stdout.
 *
 * @param raw - Raw stdout from the container.
 * @returns Extracted output text.
 */
function extractOutputText(raw: string): string {
  const stripped = raw.trim();
  if (stripped === "") return "";

  const parsed = tryParseJson(stripped) ?? regexExtractJson(stripped);
  if (parsed !== null) {
    const output = parsed["output"] ?? parsed["response"];
    if (typeof output === "string") return output;
  }

  return stripped;
}

/**
 * Try parsing the entire text as a JSON object.
 *
 * @param text - Candidate JSON string.
 * @returns Parsed record or null.
 */
function tryParseJson(text: string): Record<string, unknown> | null {
  try {
    const result: unknown = JSON.parse(text);
    if (typeof result === "object" && result !== null && !Array.isArray(result)) {
      return result as Record<string, unknown>;
    }
  } catch {
    // Not valid JSON
  }
  return null;
}

/**
 * Regex-extract the first JSON object from noisy output.
 *
 * @param text - Noisy output that may contain JSON.
 * @returns Parsed record or null.
 */
function regexExtractJson(text: string): Record<string, unknown> | null {
  const match = JSON_EXTRACT_PATTERN.exec(text);
  if (match !== null) {
    return tryParseJson(match[0]);
  }
  return null;
}

/**
 * Build an error result for a handler whose circuit is open.
 *
 * @param handler - The rejected handler name.
 * @returns AgentResult with ERROR status.
 */
function buildCircuitOpenResult(handler: string): AgentResult {
  return createAgentResult(AgentStatus.ERROR, {
    handler,
    error: `circuit open for handler '${handler}'`,
  });
}

/**
 * Record success or failure on the circuit breaker.
 *
 * @param breaker - The handler's circuit breaker.
 * @param status - The outcome status of the invocation.
 */
function recordOnBreaker(breaker: CircuitBreaker, status: AgentStatus): void {
  if (status === AgentStatus.SUCCESS) {
    breaker.recordSuccess();
  } else {
    breaker.recordFailure();
  }
}
